buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,407 @@
+     */
+    private fun getFunctionName(functionDeclaration: String): String {
+        val matches = nameRegex.findAll(functionDeclaration)
+        return matches.elementAt(1).value
+    }
+
+    /**
+     * Responsible for parsing a block of text surrounded by curly braces. It is
+     * assumed that parsing begins at the start of a curly block. Maintains the
+     * state of which structure the parser is currently in to ensure that
+     * encountered curly braces are valid.
+     */
+    class CurlyParser() {
+
+        /**
+         * All of the possible important structures that a token can be inside of.
+         * This is tracked to ensure that when a curly brace is encountered,
+         * whether it is part of the syntax or not can be determined.
+         */
+        enum class InsideState {
+
+            /**
+             * Not inside of anything.
+             */
+            NONE,
+
+            /**
+             * Inside of a string declaration (block of double quotes).
+             */
+            STRING,
+
+            /**
+             * Inside of a character declaration (block of single quotes)
+             */
+            CHARACTER,
+
+            /**
+             * Inside of a single line comment.
+             */
+            SINGLE_COMMENT,
+
+            /**
+             * Inside of a multi line comment.
+             */
+            MULTI_COMMENT
+        }
+
+        // Default state is that the parser is not within any structure.
+        private var insideState = InsideState.NONE
+
+        // Keeps track of the number of open curly braces that have been
+        // legally encountered (those that are not within any structure).
+        private var curlyCounter = 0
+
+        /**
+         * Parses the given token and updates the state. It is assumed that the
+         * tokens are provided as a sequence:
+         *
+         * for any token1, token2
+         * if parseToken(token1) is called and parseToken(token2) is called right after,
+         * then token2 comes directly after token1 in the stream of tokens.
+         *
+         * @param token The token to parse
+         * @return Whether or not the curly block has been fully parsed
+         */
+        fun parseToken(token: String): Boolean {
+            insideState = getUpdatedState(token)
+            val doneParsingCurlyBlock = curlyCounter == 0
+            return doneParsingCurlyBlock
+        }
+
+        /**
+         * Determines the updated state based on the token and current state.
+         *
+         * @param token The token to parse
+         * @return The updated state
+         */
+        private fun getUpdatedState(token: String): InsideState {
+            when (token) {
+                "\n" -> {
+                    if (insideState == InsideState.SINGLE_COMMENT) {
+                        return InsideState.NONE
+                    }
+                    return insideState
```

Comment:
If you decide to update this function to just set the insideState variable directly, you can omit all of the return statements. The ones that are currently returning something other than just "insideState" would instead just overwrite insideState with the new value.